### ISSUE TITLE
Support FFmpeg directly

### ISSUE THREAD

**Issue by trav90** [MoonchildProductions/Pale-Moon#719]
Since most of the required media code changes have been made and are now in master, I'm opening this as a tracking bug and adding it to the 27.1 milestone. Below is a list of things that need done (in no particular order) before we can ship it by default:

- [x] Add support for FFmpeg 3.x
- [x] Add a way to detect which version is available
- [x] Remove our dependency on libavformat (will make internal version checking simpler)
- [x] Fix several blocking compat issues and crashes with specific ffmpeg/libav versions
- [x] Remove GStreamer dependence for media tests
- [x] ~~Port patches listed in Issue #533~~
- [x] Port all applicable sec patches from Mozilla for our FFmpeg code

Note that our initial implementation will not work with MSE. I will hook up MSE and H.264 video after we drop GStreamer for video and have reliable HD video decoding.

**Comment by wolfbeast** [comment]
> Note that our initial implementation will not work with MSE.

Opened a new issue for hooking up the higher level functions.

**Comment by trav90** [comment]
PR #726 takes care of points 2 and 3, and partially fixes point 4. First post has been updated accordingly.

**Comment by trav90** [comment]
> > Note that our initial implementation will not work with MSE.
> 
> Opened a new issue for hooking up the higher level functions.

Turns out that this is incorrect. MSE + MP4 simply needs to be enabled by default on Linux (as it is on Windows) once this issue is fixed. See Issue #723

**Comment by trav90** [comment]
With the fixes in PR #737, point 4 above is nearly complete.

**Comment by trav90** [comment]
PR #750 should taking care of most of the blocking issues for point 4, and PR #752 takes care of point 5.

**Comment by wolfbeast** [comment]
Getting there!

**Comment by trav90** [comment]
The patches listed in point 6 above are for MSE code which (currently) does not apply to us, and a crash fix for ffvpx. Enabling ffvpx by default is going to require additional time and research to do. However not having ffvpx enabled does not prevent us from using FFmpeg instead of GStreamer for HD/H.264/MP4 video (which is the primary goal here).

So making point 6 not a blocker for this issue.

**Comment by wolfbeast** [comment]
So, I'm a little confused now. what is our current state here?
With all these patches, we're using ffmpeg (2.x) for MP4 decoding in Linux & WMF for the same in Windows, using a new metadata parser from ffmpeg on all OSes, and we're using Google's libvpx for VP8/VP9 decoding on all OSes? Something I misunderstood, or something I missed?

**Comment by trav90** [comment]
With all the patches so far, we are using ffmpeg 2.x  on Linux, WMF on Windows, libvpx for VP8/VP9 cross-platform, with libstagefright parsing metadata cross-platform.

As we previously discussed over IM, ffvpx is something we would like to use on Linux (and eventually cross-platform), however it is going to take substantially more time and research to get base support working (it will require more changes to the media-backend code).

That said, even if we do not add support for ffvpx until later, that doesn't prevent us from still using FFmpeg instead of GStreamer on Linux for MP4 decoding (which is what this issue is actually tracking).

So at this point adding support for FFmpeg 3.x should be the last hurdle before we can enable FFmpeg (and MSE + MP4) by default on Linux. Using ffvpx can happen later, and we will simply continue using libvpx until then.

Apologies for the confusion. Hopefully that better explains where we are currently at.

**Comment by trav90** [comment]
PR #772 Lays some of the groundwork for adding support for FFmpeg 3.x

**Comment by trav90** [comment]
Support for FFmpeg 3.x has landed with PR #794

**Comment by wolfbeast** [comment]
So, does this mean we have ffmpeg support completed now? On Linux anyway?

**Comment by trav90** [comment]
Yup, I'm looking at a couple sec issues now, and then we'll be ready to enable by default!

There's still a few things I want to do (simplify lib linking, couple of performance improvements, a little bit of cleanup, etc.), but all of that can be done after enabling.

**Comment by trav90** [comment]
PR #797 takes care of point 7.

**Comment by trav90** [comment]
FFmpeg support is enabled by default from PR #798!
